docs/design/0012-generic-association.md

```diff
@@ -0,0 +1,275 @@
+
+## Considered Options
+
+Since [PodPreset](https://kubernetes.io/docs/tasks/inject-data-application/podpreset/) is not available yet we are only considering here a mixed usage of a new CRD to specify the association information to be exposed plus some annotations on the top-level resources to let user specify how ECK should take over some part of the `Pod` spec.
+
+### Combine a new CRD with an annotation
+
+In this option a new CRD is introduced. It allows an Elastic stack application owner to specify for some third party applications which association data is required.
+
+#### Association specification as a new CRD
+
+```yaml
+apiVersion: association.k8s.elastic.co/v1alpha1
+kind: Association
+metadata:
+  name: monitor
+  namespace: monitor-ns # the Elastic stack applicaitons must live in this namespace
+spec:
+
+  ## targetNamespaces are the namespaces where the association information are copied and reconciled
+  targetNamespaces:
+    - ns1
+
+  ## These are the Elastic applications.
+  ## The namespace is automatically inherited from this custom resource.
+  elasticsearchName: monitor
+  kibanaName: monitor
+  # apmServerName: monitor
+  
+  ## User field lets the user specify a user or a Secret which contains a user and a password.
+  ## 2 versions are maintained:
+  ## 1. One in the target namespaces, as a part of the association information, it contains the username and the clear-text password
+  ## 2. An other one in the Elasticsearch namespace, it contains the hashed password, ready to be used by Elasticsearch
+  ## If empty then no user is generated, only certificates are copied as part of this association
+  user:
+    ## Role must already exists.
+    ## See https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-users-and-roles.html#k8s_creating_custom_roles
+    role: superuser
+    ## Optional and proposed here as a later improvement: user may specify a spec of the password.
+    ## Some sane defaults would be: 64 chars, mixed cases, no special characters and no ttl.
+    # passwordSpec:
+    #   length: 64
+    #   mixedCases: true
+    #   symbols: false
+    #   ttl: "24h"
+    # Or just reference a password
+    # secretName: secret-with-user-and-password
+```
+
+#### Automatic injection of the association data
+
+The new CRD allows creating and maintaining a copy of the information needed to establish an association. But, it might be worth to propose a mechanism to inject the variables automatically. Optionally, it would be possible to add some annotations to automatically inject the required data into the `Pods` _(in the targeted namespaces)_ :
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: my-application
+  annotations:
+    association.k8s.elastic.co/name: "monitor-ns/monitor" # namespaced name of the association CR
```

Comment:
In the original approach an association is a top level component which already allows to select which components are accessed by the remote application. That being said your idea is still relevant in that case because it would allow to have a better granularity from the application point of view.